snapcraft/internal/elf.py

```diff
@@ -86,52 +84,94 @@ def _crawl_for_path(*, soname: str, root_path: str,
+        file_type = str()
         symbols = list()  # type: List[Symbol]
-        symbols_section = subprocess.check_output([
-            'readelf', '--wide', '--dyn-syms', path]).decode()
+        output = subprocess.check_output([
+            'readelf', '--wide', '--file-header', '--dyn-syms', path])
+        readelf_lines = output.decode().split('\n')
         # Regex inspired by the regex in lintian to match entries similar to
         # number '1' from the following sample output
+        # ELF Header:
+        #  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00
+        #  Class:                             ELF64
+        #  Data:                              2's complement, little endian
+        #  Version:                           1 (current)
+        #  OS/ABI:                            UNIX - System V
+        #  ABI Version:                       0
+        #  Type:                              EXEC (Executable file)
+        #  Machine:                           Advanced Micro Devices X86-64
+        #  Version:                           0x1
+        #  Entry point address:               0x422270
+        #  Start of program headers:          64 (bytes into file)
+        #  Start of section headers:          1097096 (bytes into file)
+        #  Flags:                             0x0
+        #  Size of this header:               64 (bytes)
+        #  Size of program headers:           56 (bytes)
+        #  Number of program headers:         9
+        #  Size of section headers:           64 (bytes)
+        #  Number of section headers:         30
+        #  Section header string table index: 29
         #
         # Symbol table '.dynsym' contains 2281 entries:
         #   Num:    Value          Size Type    Bind   Vis      Ndx Name
         #     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND
         #     1: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND endgrent@GLIBC_2.2.5 (2)  # noqa
+        type_match = r'^\s+Type:\s+(?P<type>\w+) \(.*\)\Z'
+        type_regex = re.compile(type_match)
         symbol_match = (r'^\s+\d+:\s+[0-9a-f]+\s+\d+\s+\S+\s+\S+\s+\S+\s+'
                         r'(?P<section>\S+)\s+(?P<symbol>\S+).*\Z')
-        regex = re.compile(symbol_match)
-        for line in symbols_section.split('\n'):
-            m = regex.search(line)
-            if m:
-                section = m.group('section')
-                symbol = m.group('symbol')
-                try:
-                    name, version = symbol.split('@')
-                except ValueError:
-                    name = symbol
-                    version = ''
-                symbols.append(Symbol(name=name, version=version,
-                                      section=section))
-        return symbols
+        symbol_regex = re.compile(symbol_match)
+        in_symbols_section = False
+        for line in readelf_lines:
+            if line.startswith("Symbol table '.dynsym'"):
```

Comment:
Imagine this:

```python
section = elffile.get_section_by_name(b'.dynsym')
for symbol in section.iter_symbols():
   # ...
```

Reply:
I will when it works on anything but x86-64